### ISSUE TITLE
Strict mode property errors aren't inferring 'this' (or are just acting weird) TS2564

### ISSUE THREAD

**Issue by electricessence** [Microsoft/TypeScript#22500]
After TSConf, I decided to turn on 'strict' (all). :)

![image](https://user-images.githubusercontent.com/5899455/37319740-96f0561c-262d-11e8-89bf-19b16734d930.png)

https://github.com/electricessence/TypeScript.NET/blob/master/source/System.Linq/Linq.ts#L2947-L2966

2.7.2 Reports the errors but if you see the actual code (link above) there's nothing wrong with it.  In the past I commonly used `_` as an alias for `this` to naturally reduce code size and avoid scoping woes.  Should have no issue with it.

(Then installed 2.8.0 to see if the behavior was fixed.)

2.8.0 (next) Displays the errors in the image above, but WebStorm does not show the 'hinting'.

**Expected behavior:**
Should identify the type as `this` and treat it the same.

**Actual behavior:**
TS2564 is reported for every aliased instance of `this`.

**Comment by mhegazy** [comment]
Can you distill the issue a little bit. what is the error you do not expect? can you give us a simple code sample we can look at? 
for webstorm issues, i suggest posting on jetbrains issue tracker instead.

**Comment by electricessence** [comment]
```ts
readonly x:number;
constructor()
{
    const _ = this;
    _.x = 1; // Compilation error.
}
```

**Comment by electricessence** [comment]
Very commonly, `self` is also used as an alias and could manifest this problem.

**Comment by mhegazy** [comment]
`readonly` is just constructor wright, read everywhere. The idea is you get a chance to write to it at initialization time and done. 
if you are not directly using `this`, the compiler can not guarantee that this is the instance being "initialized". consider:

```ts
class Foo {
    readonly x: number;
    constructor(other: Foo) {
        this.x = 0;  // ok
        other.x = 1; // not good
    }
}
```

**Comment by electricessence** [comment]
I've converted my code to use `this.` and I can always do `const _:any = this;` but I thought it would be good to report this change and that I would expect it to be _aware_ that while in the constructor `this` as a type doesn't need to be validated against `readonly`.
